crates/rpc/src/dto/receipt.rs

```diff
@@ -74,6 +105,297 @@ impl SerializeForVersion for TxnFinalityStatus {
+            | TransactionVariant::InvokeV0(_)
+            | TransactionVariant::InvokeV1(_)
+            | TransactionVariant::InvokeV3(_) => {
+                return Err(serde_json::error::Error::custom(
+                    "expected L1Handler transaction",
+                ))
+            }
+        };
+
+        let mut serializer = serializer.serialize_struct()?;
+
+        serializer.flatten(&CommonReceiptProperties(self.0))?;
+        serializer.serialize_field("type", &"L1_HANDLER")?;
+        serializer.serialize_field("message_hash", &dto::NumAsHex::H256(&message_hash))?;
+
+        serializer.end()
+    }
+}
+impl SerializeForVersion for CommonReceiptProperties<'_> {
+    fn serialize(&self, serializer: Serializer) -> Result<serialize::Ok, serialize::Error> {
+        let mut serializer = serializer.serialize_struct()?;
+
+        serializer.serialize_field("transaction_hash", &dto::TxnHash(&self.0.transaction.hash))?;
+        serializer.serialize_field(
+            "actual_fee",
+            &FeePayment {
+                amount: &self.0.receipt.actual_fee,
+                transaction_version: &self.0.transaction.version(),
+            },
+        )?;
+        serializer.serialize_field("finality_status", &self.0.finality)?;
+        serializer.serialize_iter(
+            "messages_sent",
+            self.0.receipt.l2_to_l1_messages.len(),
+            &mut self.0.receipt.l2_to_l1_messages.iter().map(MsgToL1),
+        )?;
+        serializer.serialize_iter(
+            "events",
+            self.0.events.len(),
+            &mut self.0.events.iter().map(|e| dto::Event {
+                address: &e.from_address,
+                keys: &e.keys,
+                data: &e.data,
+            }),
+        )?;
+        serializer.serialize_field(
+            "execution_resources",
+            &ExecutionResources(&self.0.receipt.execution_resources),
+        )?;
+
+        serializer.end()
+    }
+}
+
+impl SerializeForVersion for FeePayment<'_> {
+    fn serialize(&self, serializer: Serializer) -> Result<serialize::Ok, serialize::Error> {
+        let mut serializer = serializer.serialize_struct()?;
+
+        serializer.serialize_field("amount", &dto::Felt(&self.amount.0))?;
+        serializer.serialize_field("unit", &PriceUnit(self.transaction_version))?;
+
+        serializer.end()
+    }
+}
+
+impl SerializeForVersion for MsgToL1<'_> {
+    fn serialize(&self, serializer: Serializer) -> Result<serialize::Ok, serialize::Error> {
+        let mut serializer = serializer.serialize_struct()?;
+
+        // FIXME: This clone could be elided if either the specification is fixed to
+        // represent this as an ETH_ADDRESS. Alternatively, we can amend dto::Felt to an
+        // enum over [Felt, EthAddress] but that seems worse than the clone imo.
+        let to_address = pathfinder_crypto::Felt::from_be_slice(self.0.to_address.0.as_bytes())
+            .expect("Ethereum address should fit in a felt");
```

Comment:
This is actually a bug in our code atm. Starknet accepts `to_address: Felt` - I'll make this change in a separate PR.